File: app/src/main/java/com/siavash/dualcamera/util/customviews/Toolbar.java
package com.siavash.dualcamera.util.customviews;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Build;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageButton;
import android.widget.RelativeLayout;

import com.siavash.dualcamera.R;
import com.siavash.dualcamera.util.StringUtil;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Customized toolbar
 * Created by sia on 8/26/15.
 */
public class Toolbar<T> extends RelativeLayout implements View.OnClickListener {

    @Bind(R.id.title) TextView title;
    @Bind(R.id.back_btn) ImageButton backButton;
    @Bind(R.id.action_btn) Button actionButton;

    private OnBackClickListener mBackCallback;
    private OnActionClickListener mActionCallback;

    public Toolbar(Context context) {
        super(context);
        setUp(context);
    }

    public Toolbar(Context context, AttributeSet attrs) {
        super(context, attrs);
        setUp(context);
    }

    public Toolbar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setUp(context);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP) public Toolbar(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        setUp(context);
    }

    public void setUp(Context context) {
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View view = inflater.inflate(R.layout.layout_toolbar, this, true);
        ButterKnife.bind(this, view);

        backButton.setOnClickListener(this);
    }

    public void setTitle(String title) {
        this.title.setText(title);
    }

    public void setCallback(T callback) {
        if (backButton.getVisibility() == View.VISIBLE)
            mBackCallback = (OnBackClickListener) callback;
        if (actionButton.getVisibility() == View.VISIBLE)
            mActionCallback = (OnActionClickListener) callback;
    }

    public void setRightButtonResource(int resId) {
        backButton.setImageResource(resId);
    }

    public void setRightButtonBitmap(Bitmap bitmap) {
        backButton.setImageBitmap(bitmap);
    }

    public void setActionButtonText(String text){
        actionButton.setText(text);
        actionButton.setUpFont(getContext(), StringUtil.FONT_AFSANEH);
    }

    public void setActionButtonVisibility(int visibility) {
        actionButton.setVisibility(visibility);
        setActionButtonCallback();
    }

    private void setActionButtonCallback() {
        actionButton.setOnClickListener(this);
    }

    @Override public void onClick(View v) {
        int id = v.getId();
        if (id == backButton.getId()) {
            mBackCallback.goBack();
        } else if (id == actionButton.getId()) {
            mActionCallback.doAction();
        }
    }

    public interface OnBackClickListener {
        void goBack();
    }

    public interface OnActionClickListener {
        void doAction();
    }
}

File: app/src/main/kotlin/com/siavash/dualcamera/utils/Toolbar.kt
package com.siavash.dualcamera.utils

import android.content.Context
import android.graphics.Typeface
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.widget.Toolbar
import android.util.AttributeSet
import android.widget.ImageButton
import android.widget.TextView
import com.siavash.dualcamera.R
import org.jetbrains.anko.layoutInflater
import org.jetbrains.anko.onClick
import org.jetbrains.anko.wrapContent

/**
 * Created by sia on 11/16/15.
 */
class Toolbar : Toolbar {
    private val title: TextView by bindView(R.id.toolbar_title)
    val right: ImageButton by bindView(R.id.toolbar_right)
    val left: ImageButton by bindView(R.id.toolbar_left)

    constructor(context: Context) : super(context) {
        setUp(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        setUp(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        setUp(context)
    }

    private fun setUp(context: Context) {
        context.layoutInflater.inflate(R.layout.layout_toolbar, this, true)
        setContentInsetsAbsolute(0, 0)
        minimumWidth = wrapContent
        minimumHeight = 56
    }

    fun setTitle(title: String) {
        this.title.text = title
    }

    fun setTitleStyle(typeface: Typeface) {
        title.typeface = typeface
    }

    fun defaultTitleStyle() {
        title.typeface = getFont(context, Font.AFSANEH)
    }

    fun setLeftItemVisibility(visibility: Int) {
        left.visibility = visibility
    }

    inline fun setLeftAction(crossinline function: () -> Unit) {
        if (left.visibility != VISIBLE) setLeftItemVisibility(VISIBLE)
        left.onClick {
            function()
        }
    }

    inline fun setRightAction(crossinline function: () -> Unit) {
        right.onClick {
            function()
        }
    }

    fun defaultHamburgerAction(drawer: DrawerLayout) {
        right.onClick {
            if (drawer.isDrawerOpen(GravityCompat.END)) {
                drawer.closeDrawer(GravityCompat.END)
            } else {
                drawer.openDrawer(GravityCompat.END)
            }
        }
    }
}